bokeh/util/api_crawler.py

```diff
@@ -0,0 +1,232 @@
+import ast, os, copy
+
+__all__ = ["api_crawler", "differ"]
+
+
+class APICrawler(object):
+    """
+    Api Crawler: Given a directory, it will use AST to crawl through each of the
+    File and produce AST nodes, which can then be parsed into a representation
+    of the structure of that file. The execution...
+    """
+
+    # Exclude these folders from search.
+    exclude = ("tests", "static", "sampledata", "mplexplorer")
+
+
+    # Point the crawler at this directory.
+    def __init__(self, directory):
+        self.directory = directory
+
+    def is_public(self, name):
+        # Determines whether a given function or class is public or private. Names
+        # with underscores are not allowed unless the name is "__init__"
+        return not name.startswith("_", 0, 1) or name == "__init__"
```

Comment:
Asked about this before, but didn't follow up. `__init__` is not the only public name that starts with an underscore. Arithmetic operators, for example, are public and start with an underscore. I think this code should read `not startswith('_') or (startswith('__') and endswith('__'))` (pseudo code). Or maybe we need to list double-underscore names we want to include, because maybe names like `__getattribute__` shouldn't be considered public (I'm not sure).